design/baremetal-operator/vendor-firmware-support.md

```diff
@@ -0,0 +1,151 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add API to support vendor firmware burn and config
+
+## Status
+
+provisional
+
+## Summary
+
+Add a new interface to allow burn and config vendor firmware.
+
+## Motivation
+
+It desired to burn and config vendor firmware before deploying the OS image.
+With hardware manager framework vendor can extend IPA to burn and configure its firmware.
+Burning and configuring firmware is done in the manual cleaning step. It desired to expose this
+functionality to BMO API.
+
+### Goals
+
+1. Agree on the format of the API and attributes
+2. Agree on Nvidia vendor format
+
+### Non-Goals
+
+1. Agree on others vendor’s format.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Cloud admin wants to burn and configure vendor firmware before OS deployed on bare-metal host.
+
+## Design Details
+
+### Implementation Details/Notes/Constraints
+
+
+IPA HardwareManagers allows vendors to add firmware burn and config
+in the manual cleaning stage.
+
+The ironic command to trigger the manual cleaning would be in
+the following syntax:
+
+```shell
+baremetal node clean node --clean-steps \
+[{"interface": "deploy", "step": <vendor function>, \
+   "args": {<vendor_arg1>: <vendor_value1>, \
+   <vendor_arg2>: <vendor_value2>, ...}}]
+```
+
+IPA HardwareManagers framework defines the following attributes:
+1. `step` - A function on the vendor hardware manager
+2. `argsinfo` - Arguments that can be required or optional to the step function
+3. `interface` - Should always be the deploy interface
+
+The proposal is to extend the BareMetalHost with VendorFirmware field.
+The VendorFirmware field includes the `step` name and `args` that can be required or optional.
+
+Nvidia hardware manager support `step` - update_nvidia_firmware with `args` - firmware_config and firmware_url
+
+The VendorFirmwareConfig for Nvidia should be the following:
+
+```yaml
+vendor_firmware:
+   update_nvidia_firmware:
+      firmware_config: "http://10.7.12.161/07_2022/config.yaml"
+      firmware_url: "http://10.7.12.161/fw/07_2022"
```

Comment:
I generally like this design. I think naming could use some improvement, especially considering that we also have OOB firmware updates. What I'd like to see is something like:

```yaml
firmwareUpdates:
    nvidia:
        - url: ...
          config: ...
    system:
        - url:
          config:
```

Here `system` means the future out-of-band updates (to be clear: you don't need to implement them, just to account for them added in the future). `url` is used, again, for compatibility with existing OOB clean steps.

Reply:
just to clarify we are on the same page when you wrote "out-of-band updates" you mean to update the firmware using iDRAC for example? I know that we have such support with Dell to do out-of-band updates. 
Also not sure I understand "url is used, again, for compatibility with existing OOB clean steps." what are the existing OOO cleans steps?

Reply:
Yep, that's what I mean: https://docs.openstack.org/ironic/latest/admin/drivers/redfish.html#firmware-update-using-manual-cleaning and https://docs.openstack.org/ironic/latest/admin/drivers/ilo.html#initiating-firmware-update-as-manual-clean-step